cpp/src/arrow/util/iterator_test.cc

```diff
@@ -129,11 +130,47 @@ template <typename T>
 inline Iterator<T> EmptyIt() {
   return MakeEmptyIterator<T>();
 }
-
 inline Iterator<TestInt> VectorIt(std::vector<TestInt> v) {
   return MakeVectorIterator<TestInt>(std::move(v));
 }
 
+std::function<Future<TestInt>()> AsyncVectorIt(std::vector<TestInt> v) {
+  auto index = std::make_shared<size_t>(0);
+  auto vec = std::make_shared<std::vector<TestInt>>(std::move(v));
+  return [index, vec]() -> Future<TestInt> {
+    if (*index >= vec->size()) {
+      return Future<TestInt>::MakeFinished(IterationTraits<TestInt>::End());
+    }
+    auto next = (*vec)[*index];
+    (*index)++;
+    return Future<TestInt>::MakeFinished(next);
+  };
```

Comment:
```suggestion
  size_t index = 0;
  return [index, v]() mutable -> Future<TestInt> {
    if (index >= vec.size()) {
      return Future<TestInt>::MakeFinished(IterationTraits<TestInt>::End());
    }
    return Future<TestInt>::MakeFinished(v[index++]);
  };
```

Reply:
This is good.  I changed all of the tests in iterator_test.cc to make use of captured lambda state instead of shared pointer state.